### PR TITLE
[Feature-wip](MySQL Load)Support cancel query for mysql load

### PR DESCRIPTION
**Pull request by SaintBacchus**
# Proposed changes

Issue Number: close #xxx

## Problem summary
Notice some changes:
1. Support cancel query for mysql load 
2. Change the thread pool for mysql load manager.
3. Fix sucret path check logic
4. Fix some doc error

## Checklist(Required)

* [ ] Does it affect the original behavior
* [x] Has unit tests been added
* [x] Has document been added or modified
* [ ] Is this PR support rollback (If NO, please explain WHY)
* [ ] Does it need to update dependencies

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: fe/fe-core/src/main/java/org/apache/doris/qe/StmtExecutor.java [hunk 1]
```diff
@@ -1007,6 +1008,11 @@ public void cancel() {
         if (coordRef != null) {
             coordRef.cancel();
         }
+        LOG.info("cancel1()");
```
Comment (SaintBacchus) [inline fe/fe-core/src/main/java/org/apache/doris/qe/StmtExecutor.java:1011]: Delete this

**Comment by SaintBacchus** [comment]
run buildall

**Comment by SaintBacchus** [comment]
run buildall

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 34 seconds
 stream load tsv:          485 seconds loaded 74807831229 Bytes, about 147 MB/s
 stream load json:         41 seconds loaded 2358488459 Bytes, about 54 MB/s
 stream load orc:          69 seconds loaded 1101869774 Bytes, about 15 MB/s
 stream load parquet:          29 seconds loaded 861443392 Bytes, about 28 MB/s
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230302031001_clickbench_pr_106778.html

**Comment by SaintBacchus** [comment]
run p0

**Comment by SaintBacchus** [comment]
run buildall

**Comment by SaintBacchus** [comment]
run buildall

**Comment by SaintBacchus** [comment]
run buildall

**Comment by SaintBacchus** [comment]
run p0

**Comment by SaintBacchus** [comment]
run p0

**Comment by SaintBacchus** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/load/loadv2/MysqlLoadManager.java [hunk 2]
```diff
@@ -91,10 +143,51 @@ public LoadJobRowResult executeMySqlLoadJobFromStmt(ConnectContext context, Load
                     loadResult.incSkipped(result.get("NumberFilteredRows").getAsInt());
                 }
             }
+        } catch (Throwable t) {
+            LOG.error("Execute mysql load failed", t);
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/load/loadv2/MysqlLoadManager.java:147]:
```suggestion
            LOG.warn("Execute mysql load {} failed", loadId, t);
```

File: fe/fe-core/src/main/java/org/apache/doris/load/loadv2/MysqlLoadManager.java [hunk 3]
```diff
@@ -91,10 +143,51 @@ public LoadJobRowResult executeMySqlLoadJobFromStmt(ConnectContext context, Load
...
+            // drain the data from client conn util empty packet received, otherwise the connection will be reset
+            if (loadContextMap.containsKey(loadId) && !loadContextMap.get(loadId).isFinished()) {
+                LOG.warn("not drained yet, try reading left data from client connection.");
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/load/loadv2/MysqlLoadManager.java:150]: print load id in log, or this log is meaningless

File: fe/fe-core/src/main/java/org/apache/doris/load/loadv2/MysqlLoadManager.java [hunk 4]
```diff
@@ -91,10 +143,51 @@ public LoadJobRowResult executeMySqlLoadJobFromStmt(ConnectContext context, Load
                     loadResult.incSkipped(result.get("NumberFilteredRows").getAsInt());
                 }
             }
+        } catch (Throwable t) {
+            LOG.error("Execute mysql load failed", t);
+            // drain the data from client conn util empty packet received, otherwise the connection will be reset
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/load/loadv2/MysqlLoadManager.java:148]: Do we have the same problem in normal mysql connection?

File: fe/fe-core/src/main/java/org/apache/doris/load/loadv2/MysqlLoadManager.java [hunk 5]
```diff
@@ -91,10 +143,51 @@ public LoadJobRowResult executeMySqlLoadJobFromStmt(ConnectContext context, Load
...
+            if (loadContextMap.containsKey(loadId) && !loadContextMap.get(loadId).isFinished()) {
+                LOG.warn("not drained yet, try reading left data from client connection.");
+                ByteBuffer buffer = context.getMysqlChannel().fetchOnePacket();
+                // MySql client will send an empty packet when eof
+                while (buffer != null && buffer.limit() != 0) {
+                    buffer = context.getMysqlChannel().fetchOnePacket();
+                }
+                LOG.warn("Finished reading the left bytes.");
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/load/loadv2/MysqlLoadManager.java:156]: should be debug?

File: fe/fe-core/src/main/java/org/apache/doris/load/loadv2/MysqlLoadManager.java [hunk 6]
```diff
@@ -147,7 +244,11 @@ private void fillByteBufferAsync(ConnectContext context, ByteBufferNetworkInputS
                     inputStream.fillByteBuffer(buffer);
                     buffer = context.getMysqlChannel().fetchOnePacket();
                 }
+                if (loadContextMap.containsKey(loadId)) {
+                    loadContextMap.get(loadId).setFinished(true);
+                }
             } catch (IOException | InterruptedException e) {
+                LOG.warn("Failed fetch packet from mysql client", e);
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/load/loadv2/MysqlLoadManager.java:251]: print load id

File: fe/fe-core/src/main/java/org/apache/doris/qe/StmtExecutor.java [hunk 7]
```diff
@@ -1008,6 +1009,9 @@ public void cancel() {
         if (coordRef != null) {
             coordRef.cancel();
         }
+        if (mysqlLoadId != null) {
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/qe/StmtExecutor.java:1012]: Should we reset `mysqlLoadId` after job is cancelled or finished?

File: fe/fe-common/src/main/java/org/apache/doris/common/Config.java [hunk 8]
```diff
@@ -2037,5 +2037,8 @@ public class Config extends ConfigBase {
      */
     @ConfField(mutable = false, masterOnly = false)
     public static String mysql_load_server_secure_path = "";
+
+    @ConfField(mutable = false, masterOnly = false)
+    public static int mysql_load_thread_pool = 4;
```
Comment (morningman) [inline fe/fe-common/src/main/java/org/apache/doris/common/Config.java:2042]: update the document

File: fe/fe-core/src/main/java/org/apache/doris/load/loadv2/MysqlLoadManager.java (hunk 4 above)
Comment (SaintBacchus) [reply fe/fe-core/src/main/java/org/apache/doris/load/loadv2/MysqlLoadManager.java:148]: Only mysql load will continuous send bytes. Other sql will send query at beginning

File: fe/fe-core/src/main/java/org/apache/doris/qe/StmtExecutor.java (hunk 7 above)
Comment (SaintBacchus) [reply fe/fe-core/src/main/java/org/apache/doris/qe/StmtExecutor.java:1012]: The next Stmt  in the same session will create a new `ExecuteStmt` object in class `ConnectProcessor`.
So this load id will gc after load finished and it's the reson why coordRef is not assgin null after call the cancel.

Comment (SaintBacchus) [reply fe/fe-core/src/main/java/org/apache/doris/qe/StmtExecutor.java:1012]: And in mysql Load manager, I also had check the loadId exist before use it.

**Comment by SaintBacchus** [comment]
run buildall

**Comment by SaintBacchus** [comment]
run compile

**Comment by SaintBacchus** [comment]
run  p0

**Comment by SaintBacchus** [comment]
run p0

**Comment by SaintBacchus** [comment]
run p1

**Comment by SaintBacchus** [comment]
run p1

**Review by morningman — APPROVED** [review]
LGTM

**Pull request merged**